File: Dev/C++/Mascaret/src/BEHAVE/ACLMessage.cpp
#include "BEHAVE/ACLMessage.h"
#include "BEHAVE/AID.h"

#include "Tools/Loaders/XmlParser.h"

namespace BEHAVE
{


ACLMessage::ACLMessage(ACLPerformative performative) : _performative(performative), _hasXMLContent(false)
{}

ACLMessage::~ACLMessage()
{}

string ACLMessage::getPerformativeText()
{
	string res;

	switch(_performative)
	{
	case ACCEPT_PROPOSAL : res = "ACCEPT_PROPOSAL";
	break;
	case AGREE : res = "AGREE";
	break;
	case CANCEL : res = "CANCEL";
	break;
	case CFP : res = "CFP";
	break;
	case CONFIRM : res = "CONFIRM";
	break;
	case DISCONFIRM : res = "DISCONFIRM";
	break;
	case FAILURE : res = "FAILURE";
	break;
	case INFORM : res = "INFORM";
	break;
	case INFORM_IF : res = "INFORM_IF";
	break;
	case INFORM_REF : res = "INFORM_REF";
	break;
	case NOT_UNDERSTOOD : res = "NOT_UNDERSTOOD";
	break;
	case PROPAGATE : res = "PROPAGATE";
	break;
	case PROPOSE :  res = "PROPOSE";
	break;
	case PROXY : res = "PROXY";
	break;
	case QUERY_IF : res = "QUERY_IF";
	break;
	case QUERY_REF : res = "QUERY_REF";
	break;
	case REFUSE : res = "REFUSE";
	break;
	case REJECT_PROPOSAL : res = "REJECT_PROPOSAL";
	break;
	case REQUEST : res = "REQUEST";
	break;
	case REQUEST_WHEN : res = "REQUEST_WHEN";
	break;
	case REQUEST_WHENEVER : res = "REQUEST_WHENEVER";
	break;
	case SUBSCRIBE : res = "SUBSCRIBE";
	break;
	case UNKNOWN : res = "UNKNOWN";
	default : res = "UNKNOWN";
	}
	return res;
}

string ACLMessage::toXML()
{
	string result;
	char portS[10];

	shared_ptr<XmlParser> parser = shared_ptr<XmlParser>(new XmlParser());
	parser->createFile(getPerformativeText());
	shared_ptr<XmlNode> root = parser->getRoot();

	shared_ptr<XmlNode> receiversNode = shared_ptr<XmlNode> (new XmlNode("Receivers"));
	root->addChild(receiversNode);
	for (unsigned int i = 0; i < _receivers.size(); i++)
	{
		shared_ptr<XmlNode> receiverNode = shared_ptr<XmlNode> (new XmlNode("Receiver"));
		receiversNode->addChild(receiverNode);
		receiverNode->addTextChild("Name",_receivers[i]->getName());
		receiverNode->addTextChild("Plateform",_receivers[i]->getPlateformName());
		sprintf(portS,"%d",_receivers[i]->getPlateformPort());
		receiverNode->addTextChild("Port",portS);
	}

	shared_ptr<XmlNode> senderNode = shared_ptr<XmlNode> (new XmlNode("Sender"));
	root->addChild(senderNode);
	senderNode->addTextChild("Name",_sender->getName());
	senderNode->addTextChild("Plateform",_sender->getPlateformName());
	sprintf(portS,"%d",_sender->getPlateformPort());
	senderNode->addTextChild("Port",portS);

	if (_conversationID == "") _conversationID = "not_set";
	root->addTextChild("ConversationID", _conversationID);

	if (_hasXMLContent)
	{
		shared_ptr<XmlNode> contentNode = shared_ptr<XmlNode> (new XmlNode("Content"));
		root->addChild(contentNode);
		contentNode->addChild(_xmlContent);
	}
	else
		root->addTextChild("Content", _content);

	result = parser->writeString();

	return result;
}

void ACLMessage::setXMLContent(shared_ptr<XmlNode> content)
{
	_hasXMLContent = true;
	_xmlContent = content;
}

shared_ptr<XmlNode> ACLMessage::getXMLContent(void)
{
	return _xmlContent;
}


}

File: Dev/CS/Mascaret/Mascaret/BEHAVE/ACLMessage.cs
using System;
using System.Xml.Linq;
using System.Collections.Generic;

namespace Mascaret
{

    public enum ACLPerformative { ACCEPT_PROPOSAL, AGREE, CANCEL, CFP, CONFIRM, DISCONFIRM, FAILURE, INFORM, INFORM_IF, INFORM_REF, NOT_UNDERSTOOD, PROPAGATE, PROPOSE, PROXY, QUERY_IF, QUERY_REF, REFUSE, REJECT_PROPOSAL, REQUEST, REQUEST_WHEN, REQUEST_WHENEVER, SUBSCRIBE, UNKNOWN };

    public class ACLMessage
    {
        private ACLPerformative performative;
        public ACLPerformative Performative
        {
            get { return performative; }
            set { performative = value; }
        }

        private List<AID> receivers = new List<AID>();
        public List<AID> Receivers
        {
            get { return receivers; }
            set { receivers = value; }
        }

        private List<AID> replyTo = new List<AID>();
        public List<AID> ReplyTo
        {
            get { return replyTo; }
            set { replyTo = value; }
        }

        private string content;
        public string Content
        {
            get { return content; }
            set { content = value; }
        }

        private string conversationID;
        public string ConversationID
        {
            get { return conversationID; }
            set { conversationID = value; }
        }

        private string language;
        public string Language
        {
            get { return language; }
            set { language = value; }
        }

        private string onthology;
        public string Onthology
        {
            get { return onthology; }
            set { onthology = value; }
        }

        private string protocol;
        public string Protocol
        {
            get { return protocol; }
            set { protocol = value; }
        }

        private string replyBy;
        public string ReplyBy
        {
            get { return replyBy; }
            set { replyBy = value; }
        }

        private AID sender;
        public AID Sender
        {
            get { return sender; }
            set { sender = value; }
        }

        private XElement xmlContent;
        public XElement XmlContent
        {
            get { return xmlContent; }
            set { xmlContent = value; }
        }

        private bool hasXMLContent;
        public bool HasXMLContent
        {
            get { return hasXMLContent; }
            set { hasXMLContent = value; }
        }

        private TimeExpression timestamp;
        public TimeExpression Timestamp
        {
            get { return timestamp; }
            set { timestamp = value; }
        }

        public ACLMessage(ACLPerformative performative)
        {
            this.performative = performative;
        }

        public string getPerformativeText()
        {
            string res;

            switch (performative)
            {
                case ACLPerformative.ACCEPT_PROPOSAL: res = "ACCEPT_PROPOSAL";
                    break;
                case ACLPerformative.AGREE: res = "AGREE";
                    break;
                case ACLPerformative.CANCEL: res = "CANCEL";
                    break;
                case ACLPerformative.CFP: res = "CFP";
                    break;
                case ACLPerformative.CONFIRM: res = "CONFIRM";
                    break;
                case ACLPerformative.DISCONFIRM: res = "DISCONFIRM";
                    break;
                case ACLPerformative.FAILURE: res = "FAILURE";
                    break;
                case ACLPerformative.INFORM: res = "INFORM";
                    break;
                case ACLPerformative.INFORM_IF: res = "INFORM_IF";
                    break;
                case ACLPerformative.INFORM_REF: res = "INFORM_REF";
                    break;
                case ACLPerformative.NOT_UNDERSTOOD: res = "NOT_UNDERSTOOD";
                    break;
                case ACLPerformative.PROPAGATE: res = "PROPAGATE";
                    break;
                case ACLPerformative.PROPOSE: res = "PROPOSE";
                    break;
                case ACLPerformative.PROXY: res = "PROXY";
                    break;
                case ACLPerformative.QUERY_IF: res = "QUERY_IF";
                    break;
                case ACLPerformative.QUERY_REF: res = "QUERY_REF";
                    break;
                case ACLPerformative.REFUSE: res = "REFUSE";
                    break;
                case ACLPerformative.REJECT_PROPOSAL: res = "REJECT_PROPOSAL";
                    break;
                case ACLPerformative.REQUEST: res = "REQUEST";
                    break;
                case ACLPerformative.REQUEST_WHEN: res = "REQUEST_WHEN";
                    break;
                case ACLPerformative.REQUEST_WHENEVER: res = "REQUEST_WHENEVER";
                    break;
                case ACLPerformative.SUBSCRIBE: res = "SUBSCRIBE";
                    break;
                case ACLPerformative.UNKNOWN: res = "UNKNOWN";
                    break;
                default: res = "UNKNOWN";
                    break;
            }
            return res;

        }

        public void setXMLContent(XElement contentNode)
        {
            hasXMLContent = true;
            xmlContent = contentNode;
        }

        //not implented right now
        public string toXML()
        {
            return "";
        }



    }

}
